extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -190,6 +205,74 @@ export class AmpAdExit extends AMP.BaseElement {
       this.user().error(TAG, 'Invalid JSON config', e);
       throw e;
     }
+
+    this.ampAdResourceId_ = user().assert(
+        this.getAmpAdResourceId_(),
+        `${TAG}: No friendly parent amp-ad element was found for amp-ad-exit.`);
+
+    this.unlisten_ = listen(this.getAmpDoc().win, 'message', event => {
+      const responseMessage = deserializeMessage(getData(event));
+
+      this.assertValidResponseMessage_(responseMessage, event.origin);
+
+      this.vendorResponses_[responseMessage['vendor']] =
+          responseMessage['message'];
+    });
+  }
+
+  /**
+   * Gets the resource ID of the amp-ad element containing this amp-ad-exit tag.
+   * This is a pass-through for the version in service.js, solely because
+   * the one in service.js isn't stubbable for testing, since only object
+   * methods are stubbable.
+   * @returns {?string}
+   */
+  getAmpAdResourceId_() {
+    return getAmpAdResourceId(this.element, this.win.top);
+  }
+
+  /** @override */
+  detachedCallback() {
+    if (this.unlisten_) {
+      this.unlisten_();
+    }
+  }
+
+  /**
+   *
+   * @param {?JsonObject|undefined} responseMessage The response object to validate.
```

Comment:
can this really be null|JsonObject|undefined?  just want to make sure

Reply:
Yes. It's called like this:
```
      const responseMessage = deserializeMessage(getData(event));
      this.assertValidResponseMessage_(responseMessage, event.origin);
```
And the return type of deserializeMessage() is
```
 * @returns {?JsonObject|undefined}
```
But, the first line of the method checks to make sure it's a real object.